### PR TITLE
Feature/consent adapter support

### PR DESCRIPTION
**Pull request by agnessnowplow**
- Adding missing adapter support for Databricks / Redshift / Postgres for Consent module.
- Also adding integration tests involving the consent module. So far has only been tested with Snowflake, will likely fail.

Note that for Redshift/Postgres we are introducing a dependency on [RDB Loader 4.0.0](https://github.com/snowplow/snowplow-rdb-loader/blob/b9035a5d7aa26ad4bdd1bc65d5db969ec0df8070/CHANGELOG#L149).

I'm still waiting for test data to become available for Redshift and Databricks to do a final test run on real data but I reworked the original Redshift super data for the supposedly array column within the integration test source data to be varchar according to the [schema transaltion docs](https://docs.snowplow.io/docs/understanding-tracking-design/json-schema-type-casting-rules/) already.

### CODE REVIEW

File: models/optional_modules/consent/default/snowplow_web_consent_log.sql [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  p.domain_sessionid,
+  p.derived_tstamp,
+  p.load_tstamp,
+  p.event_name,
+  p.event_type,
+  p.basis_for_processing,
+  p.consent_url,
+  p.consent_version,
+  replace(replace(replace(replace(p.consent_scopes, '"', ''), '[', ''), ']', ''), ',', ', ') as consent_scopes,
+  replace(replace(replace(replace(p.domains_applied, '"', ''), '[', ''), ']', ''), ',', ', ') as domains_applied,
```
Comment (rlh1994) [inline models/optional_modules/consent/default/snowplow_web_consent_log.sql:63]: Am I right to assume that arrays land in pg/rs as strings but wrapped with [/]'s?

File: models/optional_modules/consent/snowplow_web_consent_cmp_stats.sql [hunk 2]
```diff
@@ -5,6 +5,53 @@
   )
 }}
 
+{%- if target.type in ('postgres') -%}
```
Comment (rlh1994) [inline models/optional_modules/consent/snowplow_web_consent_cmp_stats.sql:8]: Does this also need to say redshift?

File: models/optional_modules/consent/snowplow_web_consent_cmp_stats.sql [hunk 3]
```diff
@@ -5,6 +5,53 @@
...
+  select
+    event_id,
+    domain_userid,
+    page_view_id,
+    domain_sessionid,
+    derived_tstamp,
+    event_name,
+    event_type,
+    cmp_load_time,
+    -- postgres does not allow the IGNORE NULL clause within last_value(), below workaround should do the same: removing NULLS using array_remove then using the COUNT window function (which counts the number of non-null items) to access the array using that as its index position
```
Comment (rlh1994) [inline models/optional_modules/consent/snowplow_web_consent_cmp_stats.sql:21]: Can you for my own sanity put in this comment that count is bounded up to the current row as well please, you may remember my decent in to madness before I worked that out :)

File: models/optional_modules/consent/snowplow_web_consent_scope_status.sql [hunk 4]
```diff
@@ -23,7 +23,7 @@ with arrays as (
   )
 
 select
-  replace(consent_scope,'"', '') as scope,
+  replace(replace(replace(consent_scope, '"', ''), '[', ''), ']', '') as scope,
   count(*) as total_consent
```
Comment (rlh1994) [inline models/optional_modules/consent/snowplow_web_consent_scope_status.sql:28]: What's the reaosn for this change? I assume something is returning a list (of one thing?) when it wasn't before? Feels like we're maybe doing something wrong above if this is only a single item list?

**Review by rlh1994 — COMMENTED** [review]
I haven't looked at the integration tests yet but just some questions around the new model bits

File: models/optional_modules/consent/bigquery/snowplow_web_consent_log.sql [hunk 5]
```diff
@@ -95,8 +59,8 @@ select
   p.consent_version,
   array_to_string(p.consent_scopes, ', ') as consent_scopes,
   array_to_string(p.domains_applied, ', ') as domains_applied,
-  cast(coalesce(p.gdpr_applies, false) as boolean) gdpr_applies,
-  p.elapsed_time as cmp_load_time
+  cast(coalesce(cast(p.gdpr_applies as {{ dbt.type_boolean() }}), false) as boolean) gdpr_applies,
```
[the remark this answers is not in the document]
Comment (agnessnowplow) [reply models/optional_modules/consent/bigquery/snowplow_web_consent_log.sql:62]: I definitely changed it for some reason as it was normal beforehand but I can't remember why. I'll remove it for now and see if it fails the integration tests and take it from there.

**Review by rlh1994 — COMMENTED** [review]
Okay I haven't checked the actual data itself or run it yet, but a few more questions.

File: models/optional_modules/consent/default/snowplow_web_consent_log.sql (hunk 1 above)
Comment (agnessnowplow) [reply models/optional_modules/consent/default/snowplow_web_consent_log.sql:63]: I can't be sure so for safety I have added another replace for this.

File: models/optional_modules/consent/snowplow_web_consent_cmp_stats.sql (hunk 2 above)
Comment (agnessnowplow) [reply models/optional_modules/consent/snowplow_web_consent_cmp_stats.sql:8]: No, I tested it locally with Redshift and it passed. Docs also say it accepts the ignore nulls bit: https://docs.aws.amazon.com/redshift/latest/dg/r_WF_first_value.html

File: models/optional_modules/consent/snowplow_web_consent_scope_status.sql (hunk 4 above)
Comment (agnessnowplow) [reply models/optional_modules/consent/snowplow_web_consent_scope_status.sql:28]: The {{ snowplow_utils.get_split_to_array }} macro takes the string and creates an array out of it so that we can unnest it later. Perhaps the unnest macro should take care of this cleanup.. I checked snowflake and there everything became double quoted after the unnest macro. Perhaps we did not catch this before as the media player unnests integers, not sure.

File: integration_tests/models/source/snowflake/snowplow_web_events_stg.sql [hunk 6]
```diff
@@ -1,16 +1,163 @@
...
+  select
+    *,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].basis_for_processing as basisForProcessing,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].consent_scopes as consentScopes,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].consent_url as consentUrl,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].consent_version as consentVersion,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].domains_applied as domainsApplied,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].event_type as eventType,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].gdpr_applies as gdprApplies,
+    unstruct_event_com_snowplowanalytics_snowplow_cmp_visible_1[0]:elapsed_time as elapsedTime
```
[the remark this answers is not in the document]
Comment (agnessnowplow) [reply integration_tests/models/source/snowflake/snowplow_web_events_stg.sql:21]: I would say it does not matter as it is specifically typecasted once flattened out from the object within the model. E.g.: e.unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1:consentScopes::array as consent_scopes,

File: integration_tests/models/source/snowflake/snowplow_web_events_stg.sql [hunk 7]
```diff
@@ -1,16 +1,163 @@
...
+  event_vendor,
+  event_name,
+  event_format,
+  event_version,
+  event_fingerprint,
+  true_tstamp,
+  load_tstamp,
+  contexts_com_snowplowanalytics_snowplow_web_page_1,
+  object_construct('basisForProcessing', basisForProcessing,'consentScopes', consentScopes, 'consentUrl', consentUrl, 'consentVersion', consentVersion, 'domainsApplied', domainsApplied, 'eventType', eventType, 'gdprApplies', gdprApplies) as unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1,
+  object_construct('elapsedTime', elapsedTime) as unstruct_event_com_snowplowanalytics_snowplow_cmp_visible_1
```
[the remark this answers is not in the document]
Comment (agnessnowplow) [reply integration_tests/models/source/snowflake/snowplow_web_events_stg.sql:159]: I am coalescing the data anyway in the this run table but I can change this for future proofing if the logic needs to be changed one day (coalesce(p.gdpr_applies, false) as gdpr_applies)

File: integration_tests/models/source/databricks/snowplow_web_events_stg.sql [hunk 8]
```diff
@@ -2,15 +2,161 @@
 with prep as (
 select
   *,
-  from_json(contexts_com_snowplowanalytics_snowplow_web_page_1_0_0, 'array<struct<id:string>>') as contexts_com_snowplowanalytics_snowplow_web_page_1
+  from_json(contexts_com_snowplowanalytics_snowplow_web_page_1_0_0, 'array<struct<id:string>>') as contexts_com_snowplowanalytics_snowplow_web_page_1,
+  from_json(unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1_0_0, 'array<struct<basis_for_processing:string, consent_scopes:string, consent_url:string, consent_version:string, domains_applied:string, event_type:string, gdpr_applies:string>>') as unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1,
+  from_json(unstruct_event_com_snowplowanalytics_snowplow_cmp_visible_1_0_0, 'array<struct<elapsed_time:string>>') as unstruct_event_com_snowplowanalytics_snowplow_cmp_visible_1
```
[the remark this answers is not in the document]
Comment (agnessnowplow) [reply integration_tests/models/source/databricks/snowplow_web_events_stg.sql:7]: Not sure I understand what you mean, the integration tests run correctly so I would say it is good as it is for now. Once we have real data in databricks and in case it fails the local tests I will get back to this and modify this as it would probably mean the underlying data types differ.

File: integration_tests/models/expected/snowplow_web_consent_cmp_stats_expected_stg.sql [hunk 9]
```diff
@@ -0,0 +1,13 @@
+select
+
+  event_id,
+  domain_userid,
+  cast(page_view_id as {{ dbt.type_string() }}) as page_view_id,
```
[the remark this answers is not in the document]
Comment (agnessnowplow) [reply integration_tests/models/expected/snowplow_web_consent_cmp_stats_expected_stg.sql:5]: It's one of those cases I mentioned that the seed type casts did not work so I resolved it here.

File: integration_tests/models/expected/snowplow_web_consent_totals_expected_stg.sql [hunk 10]
```diff
@@ -0,0 +1,21 @@
+select
+
+  replace(cast(consent_version as {{ dbt.type_string() }}), '.0', '') ||'.0' as consent_version,
```
[the remark this answers is not in the document]
Comment (agnessnowplow) [reply integration_tests/models/expected/snowplow_web_consent_totals_expected_stg.sql:3]: Same as above, type casting did not work in the seed directly, the sample data is sent in such a format that it has .0 at the end. I treat that field as a string but then the .0 disappears along the way as I'm casting to string. I strip out the .05 because in some data warehouses it does not behave like this and this is to make sure that we have an even finish before I readd the .05 in the end. Not sure if I explained clearly but that's what's behind. :)

File: integration_tests/models/source/databricks/snowplow_web_events_stg.sql [hunk 11]
```diff
@@ -2,15 +2,161 @@
...
 )
 
+, struct_base as (
+  select
+    *,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].basis_for_processing::STRING as basis_for_processing,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].consent_scopes::STRING as consent_scopes,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].consent_url::STRING as consent_url,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].consent_version::STRING as consent_version,
+    unstruct_event_com_snowplowanalytics_snowplow_consent_preferences_1[0].domains_applied::STRING as domains_applied,
```
[the remark this answers is not in the document]
Comment (agnessnowplow) [reply integration_tests/models/source/databricks/snowplow_web_events_stg.sql:19]: Unfortunately the docs do not say how the array type translates in various warehouses but I am assuming that for databricks the domains_applied and consent_scopes will indeed stay as an array. Will make the changes as I'm almost sure the real data would not stay as string.

**Comment by agnessnowplow** [comment]
Just a reminder to update the package reference for both the int tests and the main in the Prepare for release commit later on.

File: models/optional_modules/consent/default/snowplow_web_consent_log.sql (hunk 1 above)
Comment (agnessnowplow) [reply models/optional_modules/consent/default/snowplow_web_consent_log.sql]: Just an FYI, in Redshift it arrives like the following string: ["necessary","preferences","statistics","marketing"] I think it should be safer to leave the slash replacements just to be on the safe side for postgres.

File: models/optional_modules/consent/snowplow_web_consent_scope_status.sql (hunk 4 above)
Comment (agnessnowplow) [reply models/optional_modules/consent/snowplow_web_consent_scope_status.sql]: So the update on this was that the unnest macro was originally done with a select statement but not wrapped inside a cte. In order to do any replacements or format changes we would need to wrap this in a separate cte and another one which takes care of that. Due to backward compatibility the decision for now is to handle this afterwards in the models themselves. Another impact we have not anticipated beforehand is that Redshift creates a super data type as a result of this unnesting. If we need string replacement working with that field it is not going to be possible, it needs to be casted back to string afterwards to do that.

File: models/optional_modules/consent/default/snowplow_web_consent_log.sql (hunk 1 above)
Comment (rlh1994) [reply models/optional_modules/consent/default/snowplow_web_consent_log.sql]: As in consent scopes has the single value of `["necessary","preferences","statistics","marketing"]` as a string? or is a super type (array?) so we need to cast it as a string and then replace those things?

File: models/base/scratch/default/snowplow_web_base_events_this_run.sql [hunk 12]
```diff
@@ -143,6 +143,7 @@ with events_this_run AS (
     a.event_version,
     a.event_fingerprint,
     a.true_tstamp,
+    a.load_tstamp,
```
Comment (emielver) [inline models/base/scratch/default/snowplow_web_base_events_this_run.sql:146]: Can we safely add this?

File: models/optional_modules/consent/scratch/bigquery/snowplow_web_consent_events_this_run.sql [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  p.page_view_id,
+  p.domain_sessionid,
+  p.derived_tstamp,
+  p.load_tstamp,
+  p.event_name,
+  p.event_type,
+  p.basis_for_processing,
+  p.consent_url,
+  p.consent_version,
+  array_to_string(p.consent_scopes, ', ') as consent_scopes,
```
Comment (emielver) [inline models/optional_modules/consent/scratch/bigquery/snowplow_web_consent_events_this_run.sql:54]: Why not use `snowplow_utils.array_to_string` here as well? We're doing that in Databricks/Snowflake as well

File: models/optional_modules/consent/scratch/default/snowplow_web_consent_events_this_run.sql [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+  p.page_view_id,
+  p.domain_sessionid,
+  p.derived_tstamp,
+  p.load_tstamp,
+  p.event_name,
+  p.event_type,
+  p.basis_for_processing,
+  p.consent_url,
+  p.consent_version,
+  replace(replace(replace(replace(p.consent_scopes, '"', ''), '[', ''), ']', ''), ',', ', ') as consent_scopes,
```
Comment (emielver) [inline models/optional_modules/consent/scratch/default/snowplow_web_consent_events_this_run.sql:58]: This looks like a nightmare, is there any way we can successfully wrap this in a macro or something? So that users could leverage it themselves for custom stuff? Or do we not see that as being a common use case?

File: packages.yml [hunk 15]
```diff
@@ -1,3 +1,3 @@
 packages:
-  - package: snowplow/snowplow_utils
-    version: [">=0.13.0", "<0.14.0"]
+ - git: "https://github.com/snowplow/dbt-snowplow-utils.git"
```
Comment (emielver) [inline packages.yml:2]: This should be a separate commit imo

**Review by emielver — COMMENTED** [review]
Small changes and some open questions

File: models/base/scratch/default/snowplow_web_base_events_this_run.sql (hunk 12 above)
Comment (rlh1994) [reply models/base/scratch/default/snowplow_web_base_events_this_run.sql:146]: RBD loader 4.0.0 becomes a pre-requisite for this version so in theory yes

File: packages.yml (hunk 15 above)
Comment (rlh1994) [reply packages.yml:2]: It was, until rebase hell, if we release utils before we merge this it'll get removed and squashed.

File: models/optional_modules/consent/scratch/default/snowplow_web_consent_events_this_run.sql (hunk 14 above)
Comment (rlh1994) [reply models/optional_modules/consent/scratch/default/snowplow_web_consent_events_this_run.sql:58]: We could probably just swap it out for a `regexp_replace` call?
https://docs.aws.amazon.com/redshift/latest/dg/REGEXP_REPLACE.html

Comment (agnessnowplow) [reply models/optional_modules/consent/scratch/default/snowplow_web_consent_events_this_run.sql]: I used the translate function to replace all the individual characters in one go and then used the replace only once for adding a space after the comma. This should be clean enough.

**Comment by rlh1994** [comment]
Does this need 0.13.1 of utils or is it fine with just 0.13? @agnessnowplow

File: models/base/scratch/default/snowplow_web_base_events_this_run.sql (hunk 12 above)
Comment (rlh1994) [reply models/base/scratch/default/snowplow_web_base_events_this_run.sql:146]: @agnessnowplow Where do we document this?

File: packages.yml (hunk 15 above)
Comment (agnessnowplow) [reply packages.yml]: temporary separate commit added

File: packages.yml [hunk 16]
```diff
@@ -1,3 +1,3 @@
 packages:
   - package: snowplow/snowplow_utils
-    version: [">=0.13.0", "<0.14.0"]
+    version: [">=0.13.1", "<0.14.0"]
```
[the remark this answers is not in the document]
Comment (agnessnowplow) [reply packages.yml:3]: I understood so far that we have this practice to mark this not to forget to add this for the prepare for release commit in the release pr

[the remark this answers is not in the document]
Comment (emielver) [reply packages.yml:3]: No I don't think so. We usually have a `to delete` commit when we pin it to a temporary branch that needs to be released, but we do the bump in the initial development as well, so this can be in the main commit (unless I'm misremembering)

[the remark this answers is not in the document]
Comment (agnessnowplow) [reply packages.yml:3]: That's fine, another of those conventions that we can clarify for everyone to follow on the monthly catchup. I have changed the commit message to leave this in.

**Review by rlh1994 — APPROVED** [review]

**Review by emielver — APPROVED** [review]
👍
